[DOC-10090][DOC-10093] v24.1.0 deprecations, backward-incompatible changes, licensing updates (#18530)
cockroachdb/docs: src/current/_includes/releases/v24.1/v24.1.0.md

</div>

<h4 id="v24-1-0-enterprise-features" markdown="1">Enterprise features</h4>

{{site.data.alerts.callout_info}}
The following are [enterprise-only](https://www.cockroachlabs.com/docs/v24.1/enterprise-licensing) features. [Request a 30-day trial license](https://www.cockroachlabs.com/get-cockroachdb/enterprise/) to try them out.
{{site.data.alerts.end}}

> - The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1//licensing-faqs.md %}#obtain-a-license). Without an enterprise license, [all transactions run as `SERIALIZABLE`]({% link v24.1/transactions.md %}#aliases). [#120154](https://github.com/cockroachdb/cockroach/pull/120154)

Reviewer: Extra slash on the licensing link.

> - The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1//licensing-faqs.md %}#obtain-a-license). Otherwise, transactions which are configured to run as `READ COMMITTED` will be upgraded to [`SERIALIZABLE`]({% link v24.1/demo-serializable.md %}), as described in [Backward-incompatible changes](#read-committed-transaction-upgrade). [#120154](https://github.com/cockroachdb/cockroach/pull/120154)

Reviewer: I don't think the backward-incompatible section should be linked here. What that other note is saying is that for enterprise users on 24.1, `READ COMMITTED` will _not_ be auto-upgraded to `SERIALIZABLE` because the cluster setting to enable Read Committed is on by default. I suggested a clarification to that note as well.

```suggestion
- The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1//licensing-faqs.md %}#obtain-a-license). Without an enterprise license, [all transactions run as `SERIALIZABLE`]({% link v24.1/transactions.md %}#aliases). [#120154](https://github.com/cockroachdb/cockroach/pull/120154)
```

<h4 id="v24-1-0-backward-incompatible-changes">Backward-incompatible changes</h4>

Before [upgrading to CockroachDB v24.1]({% link v24.1/upgrade-cockroach-version.md %}), be sure to review the following backward-incompatible changes, as well as [key cluster setting changes](#v24-1-0-cluster-settings), and adjust your deployment as necessary.

- [`AS OF SYSTEM TIME`]({% link v23.2/as-of-system-time.md %}) queries can no longer use a timestamp followed by a question mark to signify a future-time value. This was an undocumented syntax. [#116830](https://github.com/cockroachdb/cockroach/pull/116830)

> - The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1//licensing-faqs.md %}#obtain-a-license). Without an enterprise license, [all transactions run as `SERIALIZABLE`]({% link v24.1/transactions.md %}#aliases). [#120154](https://github.com/cockroachdb/cockroach/pull/120154)

Reviewer: Ditto

> - The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1/licensing-faqs.md %}#obtain-a-license). Otherwise, transactions which are configured to run as `READ COMMITTED` will be upgraded to [`SERIALIZABLE`]({% link v24.1/demo-serializable.md %}), as described in the next note. [#120154](https://github.com/cockroachdb/cockroach/pull/120154)

Reviewer: Same edits as above.

```suggestion
- The [`READ COMMITTED`]({% link v24.1/read-committed.md %}) isolation level now requires the cluster to have a valid [enterprise license]({% link v24.1//licensing-faqs.md %}#obtain-a-license). Without an enterprise license, [all transactions run as `SERIALIZABLE`]({% link v24.1/transactions.md %}#aliases). [#120154](https://github.com/cockroachdb/cockroach/pull/120154)
```

> - <a id="read-committed-transaction-upgrade"></a>The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link v24.1/cluster-settings.md %}) is now `true` by default. As a result, [`READ COMMITTED`]({% link v24.1/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link v24.1/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link v24.1/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)

Reviewer: Clarifying that this only applies for enterprise users.

```suggestion
- <a id="read-committed-transaction-upgrade"></a>The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link v24.1/cluster-settings.md %}) is now `true` by default. As a result for enterprise users, [`READ COMMITTED`]({% link v24.1/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link v24.1/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link v24.1/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)
```

Before [upgrading to CockroachDB v24.1]({% link v24.1/upgrade-cockroach-version.md %}), be sure to review the following backward-incompatible changes, as well as [key cluster setting changes](#v24-1-0-cluster-settings), and adjust your deployment as necessary.

- [`AS OF SYSTEM TIME`]({% link v23.2/as-of-system-time.md %}) queries can no longer use a timestamp followed by a question mark to signify a future-time value. This was an undocumented syntax. [#116830](https://github.com/cockroachdb/cockroach/pull/116830)

> - The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link v24.1/cluster-settings.md %}) is now `true` by default. As a result, [`READ COMMITTED`]({% link v24.1/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link v24.1/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link v24.1/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)

Reviewer 2: All of the info looks correct to me. Just one more higher level question -- should this also mention that READ COMMITTED is enterprise only?

I.e., if you don't have an enterprise license, then even if the setting is on, the auto-upgrade to SERIALIZABLE will still happen. (Maybe this is something better left for the detailed docs, and I also don't want to derail this PR too much. I'm fine with whatever you decide, just mentioning it in case.)

Reviewer: OK that's an important point -- I think this needs to be a separate backward-incompatible note.

cc @mdlinville (this just keeps the wording of the release note):

```
- The `READ COMMITTED` isolation level now requires the cluster to have a valid enterprise license. [#120154](https://github.com/cockroachdb/cockroach/pull/120154)
```

Author: Done. PTAL because I added a little more detail to the end of your draft, and also added it to a new Enterprise Features section directly above the Backward-Incompatible Changes section.

> - The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-setting.md %}) is now `true` by default. As a result, [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link {{ page.version.version }}/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)

Author: ```suggestion
- The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}) is now `true` by default. As a result, [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link {{ page.version.version }}/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)
```

> - The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-setting.md %}) is now `true` by default. [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) transactions are no longer automatically upgraded to [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}), and will run as `READ COMMITTED` by default. As a result, upgrading CockroachDB may affect your application's behavior. To check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE` on v24.1, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link {{ page.version.version }}/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console. If necessary, adjust those transactions before upgrading to v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)

Reviewer: I see that you changed the `v23.2` to `v24.1`, but this is incorrect. Citing `v23.2` is intentional because the user should check whether the automatic upgrade to `SERIALIZABLE` is happening for txns on v23.2. If those txns are being issued as `READ COMMITTED`, they will not be auto-upgraded on v24.1 and the app behavior will change. This is a bit nuanced and my original version wasn't clear enough. Maybe this is better?

```suggestion
- The `sql.txn.read_committed_isolation.enabled` [cluster setting]({% link {{ page.version.version }}/cluster-setting.md %}) is now `true` by default. As a result, [`READ COMMITTED`]({% link {{ page.version.version }}/read-committed.md %}) transactions are **not** automatically upgraded to [`SERIALIZABLE`]({% link {{ page.version.version }}/demo-serializable.md %}), and will run as `READ COMMITTED` by default. On v23.2, refer to the [**Upgrades of SQL Transaction Isolation Level**]({% link {{ page.version.version }}/ui-sql-dashboard.md %}#upgrades-of-sql-transaction-isolation-level) graph in the DB Console to check whether any transaction is being upgraded from a weaker isolation level to `SERIALIZABLE`, and could therefore run differently on v24.1. [#118479](https://github.com/cockroachdb/cockroach/pull/118479)
```

Reviewer: @rafiss Would also appreciate your eyes on this!
